### PR TITLE
Add proxy address information to the metrics

### PR DESCRIPTION
**Pull request by violetagg**
Related to #3060

### CODE REVIEW

**Comment by pderop** [comment]
I missed your last commit from https://github.com/reactor/reactor-netty/pull/3081/commits/df0415ccd59a8d6bf0ccec11f919a35a7a91c8ca, I'm rechecking ...

**Comment by pderop** [comment]
Violeta, can you please clarify the following:

Prior to the last commit (https://github.com/reactor/reactor-netty/commit/df0415ccd59a8d6bf0ccec11f919a35a7a91c8ca), the AbstractChannelMetricsHandler.channelRegistered method was like this:

```
	@Override
	public void channelRegistered(ChannelHandlerContext ctx) {
		if (!onServer) {
			ProxyHandler proxyHandler = ctx.pipeline().get(ProxyHandler.class);
			if (proxyHandler != null) {
				proxyAddress = proxyHandler.proxyAddress();
			}
                         ...
```

So, after the last commit, the method is now like this:

```
	@Override
	public void channelRegistered(ChannelHandlerContext ctx) {
		if (!onServer) {
			ChannelHandler proxyHandler = ctx.pipeline().get(NettyPipeline.ProxyHandler);
			if (proxyHandler != null) {
				proxyAddress = ((ProxyHandler) proxyHandler).proxyAddress();
			}
                 ...
```
Can you please tell the reason for the last commit? 
Are there any cases where the ProxyHandler cannot be found using its class name, like it was done prior to the last commit ?

File: reactor-netty-core/src/main/java/reactor/netty/channel/ChannelMetricsRecorder.java [hunk 1]
```diff
@@ -41,13 +53,36 @@ public interface ChannelMetricsRecorder {
...
+
 	/**
 	 * Increments the number of the errors that are occurred.
 	 *
 	 * @param remoteAddress The remote peer
 	 */
 	void incrementErrorsCount(SocketAddress remoteAddress);
 
+	/**
+	 * Increments the number of the errors that are occurred.
```
[the remark this answers is not in the document]
Comment (violetagg) [reply reactor-netty-core/src/main/java/reactor/netty/channel/ChannelMetricsRecorder.java]: fixed

File: reactor-netty-core/src/main/java/reactor/netty/channel/ContextAwareChannelMetricsRecorder.java [hunk 2]
```diff
@@ -37,6 +37,18 @@ public abstract class ContextAwareChannelMetricsRecorder implements ChannelMetri
 	 */
 	public abstract void incrementErrorsCount(ContextView contextView, SocketAddress remoteAddress);
 
+	/**
+	 * Increments the number of the errors that are occurred.
```
[the remark this answers is not in the document]
Comment (violetagg) [reply reactor-netty-core/src/main/java/reactor/netty/channel/ContextAwareChannelMetricsRecorder.java]: fixed

File: reactor-netty-http/src/main/java/reactor/netty/http/client/ContextAwareHttpClientMetricsRecorder.java [hunk 3]
```diff
@@ -69,18 +102,106 @@ public abstract void recordDataSentTime(ContextView contextView, SocketAddress r
...
+	 * @param time The total time in nanoseconds for the request/response
+	 * @since 1.1.17
+	 */
+	public void recordResponseTime(ContextView contextView, SocketAddress remoteAddress, SocketAddress proxyAddress,
+			String uri, String method, String status, Duration time) {
+		recordResponseTime(contextView, remoteAddress, uri, method, status, time);
+	}
+
+	/**
+	 * Increments the number of the errors that are occurred.
```
[the remark this answers is not in the document]
Comment (violetagg) [reply reactor-netty-http/src/main/java/reactor/netty/http/client/ContextAwareHttpClientMetricsRecorder.java]: fixed

**Comment by violetagg** [comment]
> Violeta, can you please clarify the following:
> 
> Prior to the last commit ([df0415c](https://github.com/reactor/reactor-netty/commit/df0415ccd59a8d6bf0ccec11f919a35a7a91c8ca)), the AbstractChannelMetricsHandler.channelRegistered method was like this:
> 
> ```
> 	@Override
> 	public void channelRegistered(ChannelHandlerContext ctx) {
> 		if (!onServer) {
> 			ProxyHandler proxyHandler = ctx.pipeline().get(ProxyHandler.class);
> 			if (proxyHandler != null) {
> 				proxyAddress = proxyHandler.proxyAddress();
> 			}
>                          ...
> ```
> 
> So, after the last commit, the method is now like this:
> 
> ```
> 	@Override
> 	public void channelRegistered(ChannelHandlerContext ctx) {
> 		if (!onServer) {
> 			ChannelHandler proxyHandler = ctx.pipeline().get(NettyPipeline.ProxyHandler);
> 			if (proxyHandler != null) {
> 				proxyAddress = ((ProxyHandler) proxyHandler).proxyAddress();
> 			}
>                  ...
> ```
> 
> Can you please tell the reason for the last commit? Are there any cases where the ProxyHandler cannot be found using the ProxyHandler.class, like it was done prior to the last commit ?

I decided that it will be better to search for the `ProxyHandler` added by Reactor Netty and not just for some `ProxyHandler` in the pipeline.

**Comment by violetagg** [comment]
@pderop Thanks for the review!

**Pull request merged**